File: vipimo/mzunguko_mchezo/bench.go
package main

import "fmt"

type Mfumo struct {
	thamani int32
	kiwango int32
}

func mfumoUnda(mwanzo int32) *Mfumo { return &Mfumo{thamani: mwanzo, kiwango: 1} }
func (m *Mfumo) pataThamani() int32 { return m.thamani }
func (m *Mfumo) pataKiwango() int32 { return m.kiwango }
func (m *Mfumo) sasisha(pembejeo int32) {
	sasa := m.pataThamani()
	kw := m.pataKiwango()
	mpya := sasa + (pembejeo % 7) - (kw % 3)
	if mpya < 0 {
		mpya = -mpya
	}
	m.thamani = mpya
	if mpya%500 == 0 {
		m.kiwango = kw + 1
	}
}

type Mchezo struct {
	m [12]*Mfumo
}

func mchezoUnda() *Mchezo {
	starts := [12]int32{10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120}
	g := &Mchezo{}
	for i, s := range starts {
		g.m[i] = mfumoUnda(s)
	}
	return g
}

func (g *Mchezo) sasisha(zamu int32) {
	for i := 0; i < 12; i++ {
		g.m[i].sasisha(zamu + int32(i) + 1)
	}
}

const zamuIdadi = 200000

func main() {
	g := mchezoUnda()
	var z int32 = 0
	for z < zamuIdadi {
		g.sasisha(z)
		z++
	}
	jumla := g.m[0].pataThamani() + g.m[5].pataThamani() + g.m[11].pataThamani()
	fmt.Printf("jumla=%d z=%d\n", jumla, z)
}

File: vipimo/mzunguko_mchezo/bench.rs
struct Mfumo { thamani: i32, kiwango: i32 }

impl Mfumo {
    fn new(mwanzo: i32) -> Box<Mfumo> { Box::new(Mfumo { thamani: mwanzo, kiwango: 1 }) }
    fn pata_thamani(&self) -> i32 { self.thamani }
    fn pata_kiwango(&self) -> i32 { self.kiwango }
    fn sasisha(&mut self, pembejeo: i32) {
        let sasa = self.pata_thamani();
        let kw = self.pata_kiwango();
        let mut mpya = sasa + (pembejeo % 7) - (kw % 3);
        if mpya < 0 { mpya = -mpya; }
        self.thamani = mpya;
        if mpya % 500 == 0 { self.kiwango = kw + 1; }
    }
}

struct Mchezo { m: Vec<Box<Mfumo>> }

impl Mchezo {
    fn new() -> Mchezo {
        let starts = [10,20,30,40,50,60,70,80,90,100,110,120];
        Mchezo { m: starts.iter().map(|&s| Mfumo::new(s)).collect() }
    }
    fn sasisha(&mut self, zamu: i32) {
        for i in 0..12 {
            self.m[i].sasisha(zamu + i as i32 + 1);
        }
    }
}

const ZAMU_IDADI: i32 = 200000;

fn main() {
    let mut g = Mchezo::new();
    let mut z = 0;
    while z < ZAMU_IDADI {
        g.sasisha(z);
        z += 1;
    }
    let jumla = g.m[0].pata_thamani() + g.m[5].pata_thamani() + g.m[11].pata_thamani();
    println!("jumla={} z={}", jumla, z);
}
